Publish to Single-File

swaroop-sridhar opened the pull request:
This change implements support for publishing apps to a [single file](https://github.com/dotnet/designs/blob/master/accepted/single-file/design.md).

* ``dotnet publish /p:PublishSingleFile=true`` causes the contents of the "original" publish directory to a single file in the actual publish directory
* Files marked with the meta-data ``<ExcludeFromSingleFile>false<ExcludeFromSingleFile>`` are left in the publish directory unbundled. This includes PDB files by default
* PDB files can be bundled into the single file by setting ``/p:IncludeSymbolsInSingleFile=true``

Publishing to a single file requires publishing wrt a RID using an apphost, because the generated file is the platform-specific AppHost executable with embedded dependencies.

swaroop-sridhar committed:
Publish to Single-File

This change implements support for publishing apps to a [single file](https://github.com/dotnet/designs/blob/master/accepted/single-file/design.md).

* ``dotnet publish /p:PublishSingleFile=true`` causes the contents of the "original" publish directory to a single file in the actual publish directory
* Files marked with the meta-data ``<ExcludeFromSingleFile>false<ExcludeFromSingleFile>`` are left in the publish directory unbundled. This includes PDB files by default
* PDB files can be bundled into the single file by setting ``/p:IncludeSymbolsInSingleFile=true``

Publishing to a single file requires publishing wrt a RID using an apphost, because the generated file is the platform-specific AppHost executable with embedded dependencies.

swaroop-sridhar commented:
@nguerrera @fadimounir  @sbomer  please review.
CC: @ericstj

swaroop-sridhar commented:
@dotnet-bot  test this please

sbomer commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -85,6 +86,9 @@ Copyright (c) .NET Foundation. All rights reserved.
     -->
   <Target Name="PrepareForPublish">
 
+    <NETSdkError Condition="'$(_IsExecutable)' != 'true'"
```
Oh you already fixed it. :)

swaroop-sridhar replied:
Thanks @sbomer; as you know, I saw the feedback over your shoulder :)

swaroop-sridhar commented:
Thanks @sbomer. Once the ILLinker change (https://github.com/dotnet/sdk/pull/3125) goes through, I'll add another test for Illink + R2R + Bundled app.

fadimounir commented on src/Tasks/Common/Resources/Strings.resx:
```diff
@@ -505,4 +505,16 @@ The following are names of parameters or literal values and should not be transl
+    <comment>{StrBegin="NETSDK1098: "}</comment>
+  </data>
+  <data name="CannotHaveSingleFileWithoutExecutable" xml:space="preserve">
+    <value>NETSDK1099: Publishing to a single-file is currently only supported for executable applications.</value>
```
>currently [](start = 54, length = 9)

I feel we should remove "currently". If we support this in the future, we can delete this error message

fadimounir commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -770,6 +792,47 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <PropertyGroup>
+      <!-- Handle the case where IncludeSymbolsInSingleFile may not be defined -->
+      <EmbedPDBs Condition="'$(IncludeSymbolsInSingleFile)' == 'true'">true</EmbedPDBs>
+      <EmbedPDBs Condition="'$(IncludeSymbolsInSingleFile)' != 'true'">false</EmbedPDBs>
```
We should also handle the .pdb/.map files created by crossgen here. We can do this later (I want to make some changes to the R2R related stuff first). Please file an issue to track this.

swaroop-sridhar replied:
https://github.com/dotnet/core-setup/issues/5903

swaroop-sridhar committed:
Update an error message.

swaroop-sridhar committed:
Change the R2R property in a test

The property to trigger R2R compilation was changed from "Ready2Run" to "PublishReadyToRun".
Update this in a test case.

swaroop-sridhar commented:
In this change https://github.com/dotnet/sdk/pull/3132/commits/1cc93522a00a961281242ded2e629a26158e6e83 @fadimounir I updated  a test case to use the new name `PublishReadyToRun` for triggering R2R compilation, now that https://github.com/dotnet/sdk/pull/3134 is checked in.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs:
```diff
@@ -0,0 +1,37 @@
+            foreach(var item in FilesToBundle)
+            {
+                fileSpec.Add(new FileSpec(sourcePath: item.ItemSpec,
+                                          bundleRelativePath: item.GetMetadata("RelativePath")));
```
```suggestion
                                          bundleRelativePath: item.GetMetadata(MetadataKeys.RelativePath)));
```

Not sure if I got that exactly right, but there's a class with constants for metadata. Use that.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs:
```diff
@@ -0,0 +1,37 @@
+            Bundler bundler = new Bundler(AppHostName, OutputDir, EmbedPDBs, ShowDisgnosticOutput);
+            var fileSpec = new List<FileSpec>(FilesToBundle.Length);
+
+            foreach(var item in FilesToBundle)
```
```suggestion
            foreach (var item in FilesToBundle)
```
nit: spacing

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -121,20 +127,28 @@ Copyright (c) .NET Foundation. All rights reserved.
           Outputs="@(_ResolvedFileToPublishPreserveNewest->'$(PublishDir)%(RelativePath)')">
 
+    <ItemGroup>
+      <_ResolvedUnBundledFileToPublishPreserveNewest 
```
```suggestion
      <_ResolvedUnbundledFileToPublishPreserveNewest 
```

nit on casing. Un is not a word on its own.

nguerrera replied:
(Obviously, rename all references to it too.)

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -719,6 +741,47 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <PropertyGroup>
+      <!-- Handle the case where IncludeSymbolsInSingleFile may not be defined -->
+      <EmbedPDBs Condition="'$(IncludeSymbolsInSingleFile)' == 'true'">true</EmbedPDBs>
+      <EmbedPDBs Condition="'$(IncludeSymbolsInSingleFile)' != 'true'">false</EmbedPDBs>
```
EmbedPDBs is a pretty generic name since there is also a concept of embeddign PDB in dll (DebugType=Embedded). You can avoid having a new name, by defaulting IncludeSymbolsInSingleFile to false when empty. This is very idiomatic in msbuild, you will see it all over the place.

``` xml
<PropertyGroup>
   <IncludeSymbolsInSingleFile Condition="'$(IncludeSymbolsInSingleFile)' == ''">false</IncludeSymbolsInSingleFile>
</PropertyGroup>
```
I don't even think you need the comment if you do that, it is pretty obvious what this handles.

Then pass EmbedPdbs="$(IncludeSymbolsInSingleFile)

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs:
```diff
@@ -0,0 +1,37 @@
+        [Required]
+        public string AppHostName { get; set; }
+        [Required]
+        public bool EmbedPDBs { get; set; }
```
```suggestion
        public bool IncludeSymbols { get; set; }
```

I think it is worth using the same terminology throughout, and the public facing name is symbols for this feature.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs:
```diff
@@ -0,0 +1,37 @@
+
+        protected override void ExecuteCore()
+        {
+            Bundler bundler = new Bundler(AppHostName, OutputDir, EmbedPDBs, ShowDisgnosticOutput);
```
```suggestion
            var bundler = new Bundler(AppHostName, OutputDir, EmbedPDBs, ShowDisgnosticOutput);
```

nit: type is obvious

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -719,6 +741,47 @@ Copyright (c) .NET Foundation. All rights reserved.
+          Condition="'$(PublishSingleFile)' == 'true'"
+          DependsOnTargets="_ComputeFilesToBundle" 
+          Inputs="@(_FilesToBundle)"
+          Outputs="$(PublishDir)\$(AssemblyName)$(_NativeExecutableExtension)">
```
```suggestion
          Outputs="$(PublishDir)$(AssemblyName)$(_NativeExecutableExtension)">
```

We ensure trailing slash and you can see this is how PublishDir is used elsewhere. This is another common idiom in msbuild, it prevents accidental writing to root of disk if something is undefined. I once deleted my entire hard drive this way so I stick to this now :)

ericstj replied:
Funny story, a similar issue caused CoreFx build perf to tank once, the cause `$(SomeProperty)\**\*.extension`.  Someone renamed `SomeProperty` and CI passed...

nguerrera reviewed (commented):
Looks good overall. Some minor feedback.

swaroop-sridhar committed:
Address feedback by @nguerrera

swaroop-sridhar commented:
@nguerrera: I've made the corrections you suggested in this commit: 9246cad 
Thanks.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs:
```diff
@@ -0,0 +1,37 @@
+            var bundler = new Bundler(AppHostName, OutputDir, IncludeSymbols, ShowDisgnosticOutput);
+            var fileSpec = new List<FileSpec>(FilesToBundle.Length);
+
+            Array.ForEach(FilesToBundle, 
```
Why did you change this to Array.ForEach? I liked the foreach loop better.

swaroop-sridhar replied:
OK @nguerrera, I pushed 25c2b423608690a37da83dbdcecdc55138af67cb which revers back to the foreach() loop.

swaroop-sridhar committed:
Change an Array.Foreach() to foreach() loop.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs:
```diff
@@ -0,0 +1,38 @@
+        [Required]
+        public bool ShowDisgnosticOutput { get; set; }
+
+        protected override void ExecuteCore()
```
I love how delightfully simple this task ended up being. :)

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs:
```diff
@@ -0,0 +1,38 @@
+        [Required]
+        public string OutputDir { get; set; }
+        [Required]
+        public bool ShowDisgnosticOutput { get; set; }
```
```suggestion
        public bool ShowDiagnosticOutput { get; set; }
```

Just spotted a typo

swaroop-sridhar replied:
Thanks, fixed it in 77e4d367b557e788242cee0b3406b83b298b5b0a

swaroop-sridhar committed:
Fix a typo in a variablename

Thanks @nguerrera

swaroop-sridhar commented:
@dotnet-bot  test dotnet-sdk-public-ci

ericstj reviewed (approved):
Nice!

cup commented:
https://github.com/dotnet/cli/issues/7737
https://github.com/dotnet/core/issues/1278
https://github.com/dotnet/core/issues/2647
https://github.com/dotnet/coreclr/issues/17015
https://github.com/dotnet/coreclr/issues/20287
https://github.com/dotnet/corefx/issues/13329
https://github.com/dotnet/corert/issues/6819
https://github.com/dotnet/core-setup/pull/5286
https://github.com/dotnet/designs/pull/52
https://github.com/dotnet/sdk/pull/3074

https://github.com/dotnet/sdk/pull/3132
